settings: Fix invalid date flash and incorrect reset.

Tigas143 opened the pull request:
Previously, when manually entering a date in the "Birthday" custom profile field without using the date picker, a validation error briefly flashed despite the date being valid. This occurred due to two POST requests being sent, with one still containing the date in a human-readable format, which caused a 400 Bad Request. This problem is caused by flatpickr behavior and I think the best approach to solve this on Zulip's side is to only take into consideration the request that comes in the internal format.

To fix this, this PR adds a validation step to discard the outdated request that caused the invalid date flash.

Additionally, if an invalid date was entered, the field would reset to the last date selected via the date picker, even if it had been manually replaced with a valid date. While the open issue suggests this only happened when selecting a date via the picker, the current implementation actually cleared the field when an invalid date was entered manually. I believe the issue is again related to flatpickr's behavior and the best approach is to override the default flapickr way of dealing with invalid input, which is to send an empty string.

To fix this, this PR adds a `lastValidDate` variable to track the last confirmed valid date, ensuring that only valid dates persist and invalid entries are properly cleared.

A test (`initialize_custom_date_type_fields_keyboard_validation`) was added to verify that invalid manual inputs do not overwrite a previously valid date.

A test (`update_custom_profile_field`) was added to verify that POST requests of id 5 (Birthday field id) are discarded and not sent.

<!-- Describe your pull request here.-->

Fixes: #19936


<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

**Screenshots and screen captures:**

https://github.com/user-attachments/assets/f87c817c-3e0b-410c-9cfb-6825dc87505f


<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [X] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [X] Explains differences from previous plans (e.g., issue description).
- [X] Highlights technical choices and bugs encountered.
- [X] Calls out remaining decisions and concerns.
- [X] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [X] Each commit is a coherent idea.
- [X] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [X] Visual appearance of the changes.
- [X] Responsiveness and internationalization.
- [X] Strings and tooltips.
- [X] End-to-end functionality of buttons, interactions and flows.
- [X] Corner cases, error conditions, and easily imagined bugs.
</details>

Tigas143 commented:
Hi @alya,

Could you take a look at my PR (#34193)? I'm having trouble understanding the CI failures. When I run ./tools/lint or Prettier locally, I don't get any warnings about the file in question.

Tigas143 commented:
Hi @alya I think this is ready for review, could you take a look at my PR? Thanks!

Tigas143 commented:
Thank you for the feedback @alya! I've updated the commit message to match the style, could you take a look? Thanks!

Tigas143 commented:
I mentioned "Birthday" in the commit message because it was the concrete issue I was fixing. But I've updated the message to use "custom profile field" because the fix applies to any custom profile field with a date and "Birthday" was an example field.  Could you take another look @alya? Thanks!

alya commented:
@sahil839 Could you please take a look at this one? I haven't tested.

sahil839 commented on web/src/settings_account.ts:
```diff
@@ -202,14 +202,25 @@ function settings_change_error(message_html: string, xhr?: JQuery.jqXHR): void {
         data = JSON.stringify([field.id]);
     } else {
+        if (
+            field.id === 5 &&
```
This is just the field ID in development environment not everytime, so this is incorrect.

Tigas143 replied:
You're right, using field ID 5 isn't reliable outside dev. Since this fix isn’t ideal long-term, should I look into a proper way to detect date fields, or is this not worth keeping at all?

Tigas143 replied:
I believe this issue is now resolved.

sahil839 commented on web/src/settings_account.ts:
```diff
@@ -202,14 +202,25 @@ function settings_change_error(message_html: string, xhr?: JQuery.jqXHR): void {
+        if (
+            field.id === 5 &&
+            typeof field.value === "string" &&
+            !is_valid_date_format(field.value)
```
I don't get why this is needed. The problem currently is that the date should be passed in correct format when typing the value in input. I am not sure what formats to support, but we should just convert the typed format to a format accepted by API.

@alya thoughts about supported formats when typing manually?

alya replied:
I dunno, is there a standard way/libary to manage this? I feel like we aren't the only ones who need to validate typed dates.

Tigas143 replied:
This validation is needed because flatpickr handles date conversion internally, but still sends the human-readable format (e.g. “March 20, 2025”) to the API instead of the expected internal format, which can cause rejections and trigger the invalid flash error. The check ensures the value is valid before sending it. A better long-term fix would be using a library that properly validates manual input, but that would require broader changes to how date inputs are handled overall.

Tigas143 replied:
I believe this issue is now resolved.

sahil839 commented on web/src/custom_profile_fields_ui.ts:
```diff
@@ -173,11 +173,63 @@ export function initialize_custom_date_type_fields(element_id: string): void {
+        });
+        // Store lastValidDate so it remains after reopening settings
+        if (lastValidDate && Number.isNaN(lastValidDate.getTime())) {
+            $element.data("lastValidDate", lastValidDate);
```
I don't think this is needed. Simple fix should be to just add a error handler which should reset the value to original value.

Tigas143 replied:
I understand, but I’m not sure how an error handler would work in this case. Flatpickr doesn’t throw an error when an invalid date is entered, so the only reliable way I’ve found to determine when to reset the value is by explicitly checking if the date is valid. This also seems to be a limitation of the library, there’s no clear or consistent way to distinguish between valid and invalid conversions. That’s why I went with this approach. Open to suggestions if there’s a cleaner or more robust way to handle it!

sahil839 replied:
Eventually what we are doing to send the API request is calling the `.on("change"` handler [here](https://github.com/zulip/zulip/blob/5814ac559f77eece7d88e93392ec8334c689bdec/web/src/settings_account.ts#L752) . There are two inputs one is hidden and one is visible to the user. The hidden input stores the value in valid format which is passed to the API like "YYYY-MM-DD" while the visible input stores it in format as shown to the user.

When selecting the date from datepicker, the change event is being triggered on the hidden input so the correct value is passed in the API request.

When typing date in the user visible input, the change handler is called for the visible input and the value passed to the API request is just the text in the input. So, what we should change is to just update the handler to validate and convert the input to valid input if it is called due to change in user visible input.

sahil839 commented on web/src/custom_profile_fields_ui.ts:
```diff
@@ -173,11 +173,62 @@ export function initialize_custom_date_type_fields(element_id: string): void {
+                    } else {
+                        // Otherwise, clear the field
+                        instance.clear();
+                        $element.val("").trigger("change");
```
I think this can be removed now.

Tigas143 replied:
You're right, this can be removed now.

Tigas143 replied:
This has been removed.

sahil839 commented on web/src/settings_account.ts:
```diff
@@ -202,7 +202,23 @@ function settings_change_error(message_html: string, xhr?: JQuery.jqXHR): void {
+}
+function parse_to_iso_date_format(dateStr: string): string | null {
+    const parsed = new Date(dateStr);
+    if (Number.isNaN(parsed.getTime())) {
```
Is this check to handle invalid strings?

Tigas143 replied:
Yes, but it might not be necessary if the conversion is always successful. On the other hand, if we decide to handle all invalid dates, would it be correct to attempt converting any invalid inputs?

sahil839 commented on web/src/settings_account.ts:
```diff
@@ -751,12 +767,20 @@ export function set_up(): void {
+            if (converted) {
+                value = converted;
+            } else {
+                value = "";
```
I guess this is for handling invalid strings, but I don't think it is correct as it will remove the value. Value should only be removed if the text input is made empty, but not when input having invalid strings.

I am not sure how to handle invalid strings here, there isn't any button which is clicked to confirm changes which can be disabled. @alya thoughts?

sahil839 replied:
> I am not sure how to handle invalid strings here, there isn't any button which is clicked to confirm changes which can be disabled. @alya thoughts?

@alya any feedback on how to handle invalid strings?

alya replied:
Oh, sorry to miss the question earlier. We should probably outline the field in red, and revert to the previous value when it loses focus.

sahil839 commented on web/src/settings_account.ts:
```diff
@@ -751,12 +767,20 @@ export function set_up(): void {
             $(this).closest(".custom_user_field").attr("data-field-id")!,
             10,
         );
+        if ($(this).hasClass("date-field") && !is_valid_date_format(value)) {
```
I think it is better to not have this `is_valid_date_format` and just validate the string using `Date` itself for all cases. Because a string like "2025-14-13" will be invalid date as "14" is not a valid month value.

Tigas143 replied:
I understand, but I still think `is_valid_date_format` is necessary because `Date()` treats the visible format as valid. Would it be the correct approach to use both validations? That way, something like "2025-14-13" would be rejected, even if it probably can’t reach the API through manual input.

sahil839 replied:
I don't get what do you mean by visible format. I guess you are talking about strings like "April 10, 2025". I guess typing that should be fine and `Date` can take both formats including "2025-04-10" as inputs. We only need to make sure while sending the API request, the format should be "2025-04-10".

Also, a point to note is that `Date` will not parse something like "10-04-2025" as 10th April 2025, instead it will be parsed as 4th October 2025.

We can wait for @alya feedback for now on how to handle invalid strings, so that we get clear idea of how to proceed.

Tigas143 replied:
Hi @sahil839, I've removed `is_valid_date_format` and now only check if the conversion can be done using `Date`. If it fails, the value remains unchanged. Let me know if this approach looks good to you.

Tigas143 commented:
Hi @alya and @sahil839, I’d like to move forward with this. I’m happy to make any changes needed regarding how to handle invalid strings. Thanks!

Tigas143 commented:
Thanks for the feedback, @sahil839. Could you please clarify which part you believe is unnecessary? The changes in `custom_profile_fields_ui` are required to reset the date and to identify the field using the "date-field" class. I'm not quite sure what exactly you're suggesting should be removed.
Also, could you clarify what you meant by "code to handle invalid strings"? I'm no longer modifying the `value` variable, there's just an attempt to convert it into a Date.

sahil839 commented:
This part -
```js
$date_picker_elements.each((_, element) => {
        const $element = $(element);
        const storedDate: unknown = $element.data("lastValidDate");
        let lastValidDate: Date | undefined;
        if (storedDate instanceof Date) {
            lastValidDate = storedDate;
        }
        if (!$element.hasClass("date-field")) {
            $element.addClass("date-field");
        }
        flatpickr(element, {
            altInput: true,
            altFormat: "F j, Y",
            allowInput: true,
            static: true,
            onReady(_, __, instance) {
                const rawValue = $element.val();
                const initialValue = Array.isArray(rawValue) ? rawValue[0] : rawValue;
                // Parse and set initial date if a valid string is provided
                if (typeof initialValue === "string" && initialValue.trim()) {
                    const parsedDate = instance.parseDate(initialValue, instance.config.dateFormat);
                    if (parsedDate) {
                        lastValidDate = parsedDate;
                        instance.setDate(parsedDate, true);
                    }
                }
            },
            onChange(selectedDates, _, instance) {
                if (selectedDates.length > 0) {
                    lastValidDate = selectedDates[0];
                    if (lastValidDate !== undefined) {
                        // Save the date as the lastValidDate
                        $element.data("lastValidDate", lastValidDate);
                        $element
                            .val(instance.formatDate(lastValidDate, instance.config.dateFormat))
                            .trigger("change");
                    }
                }
            },
            onClose(_, dateStr, instance) {
                const parsedDate = instance.parseDate(dateStr, instance.config.dateFormat);
                if (!parsedDate) {
                    if (lastValidDate) {
                        // Restore last valid date if available
                        instance.setDate(lastValidDate, true);
                        $element
                            .val(instance.formatDate(lastValidDate, instance.config.dateFormat))
                            .trigger("change");
                    } else {
                        // Otherwise, clear the field
                        instance.clear();
                        $element.val("").trigger("change");
                    }
                }
            },
        });
```

Tigas143 commented:
Hi @sahil839, I believe I’ve removed the unnecessary changes and cleaned up the commit history. The reset is now handled in `settings_account.ts`. However, I’m running into an issue where invalid dates (e.g., "bananas") are being passed as empty strings. Based on my `console.log` tests, this appears to be caused by Flatpickr. Since the invalid strings are being silently converted to empty strings, I’m unable to handle them properly. 

I’d appreciate any feedback, and please let me know if there are any other changes you'd like me to make. Thanks!

![Screenshot_721](https://github.com/user-attachments/assets/f71caf0f-4bc9-4f89-b479-57a9211428e9)

Tigas143 commented:
Hi @sahil839, just wanted to follow up on my previous message in case it got missed. Let me know if you have any feedback or if there’s anything I should adjust. Thanks again!

sahil839 commented:
- I think the behavior for invalid strings is not implemented yet. @alya mentioned about showing red outline, but am not sure whether we want to validate the string when typing each character. I was thinking if we can just show an error message at the place where "Saving" indicator is shown correctly which will automatically hide after some time like 2 seconds and reset the input to original value and this will be done once the input loses focus. @alya thoughts?

- The input is again set to original value when input is cleared which is not the expected behavior. The behavior should be similar to what we do when clicking on "x".

sahil839 commented:
I think implementation can be simplified and the code can be made easy to read and understand. A couple of points-
- To set the input to original value, there is no need to store the value in `data("lastValidValue")`. `people.get_custom_profile_data` function can be used.
- Also, I think a cleaner solution would be to just add a separate handler for this input which will just handle updating the value when typing in the input manually. You can use `altInputClass` config property for having a custom class for this input which can be used for "change" event handler. And also `flatpickr_instance.setDate` can be used to keep the code clean.

Tigas143 commented:
Hi @sahil839, thank you for the feedback I am working on simplifying and cleaning up the code.



> * I think the behavior for invalid strings is not implemented yet.

It's not implemented because Im not sure what direction to take. As I mentioned in my previous message,  currently the invalid strings are not being passed, instead the `value` becomes an empty string (`""`).




> * The input is again set to original value when input is cleared which is not the expected behavior. The behavior should be similar to what we do when clicking on "x".

The root of this behavior is that I cant distinguish invalid input from empty input in the "change" event handler, `value` will be an empty string in both cases.

I would really appreciate any specific feedback or suggestions on how to approach this, as I’m still trying to understand why the current behavior exists.

alya commented:
> I think the behavior for invalid strings is not implemented yet. @alya mentioned about showing red outline, but am not sure whether we want to validate the string when typing each character. I was thinking if we can just show an error message at the place where "Saving" indicator is shown correctly which will automatically hide after some time like 2 seconds and reset the input to original value and this will be done once the input loses focus. @alya thoughts?

Sure, that's fine. This is not a high-visibility feature, so we should aim to get to a reasonable UX without over-complicating the implementation.

sahil839 commented:
Invalid strings can be handled by setting `dateParser` config option for flatpicker to `new Date(date_string)`. And empty string can simply be handled separately by checking the input value.

Tigas143 commented:
Hi @sahil839 I have implemented the behavior for invalid strings, fixed the empty string behavior and show an error message when the date input is invalid. Let me know if you have any feedback or if there’s anything I should adjust. Thanks!

timabbott commented:
Marking as a completion candidate. I don't think maintainers can spend any more time on this PR given how much effort it's taking to explain the changes required; it's requiring too much handholding and still isn't near the readability goals we have for PRs we merge.

sahil839 commented:
Closing this as the issue was fixed in #34956. Thanks for all your work on this @Tigas143.
